### ISSUE TITLE
Provide a bookmarking system like in Kiwix Desktop

### ISSUE THREAD

**Issue by Jaifroid** [kiwix/kiwix-js#757]
As suggested by @knowledge-is-power [here](https://github.com/openzim/mwoffliner/issues/1474#issuecomment-905228503), it would be good to have a bookmark system in Kiwix JS and Kiwix JS PWA/Windows. It would require #280 or similar, and probably also depends on #656 to allow seamless switching between bookmarks in different ZIM archives in a folder (to which the user has given read permission).

**Comment by kelson42** [comment]
This should simply work through the Web Browser bookmarking system. Would be a bad idea IMO to redo something special here.

**Comment by Jaifroid** [comment]
> This should simply work through the Web Browser bookmarking system. Would be a bad idea IMO to redo something special here.

Agreed in principle, but unfortunately it doesn't work even in Service Worker mode. So we need to make it work, probably with #280. Even so, that would only work in Service Worker mode, which Firefox browser-extension users can't access. Plus there are other contexts where the browser's native bookmarking system is not available: PWA, mobile. Another solution is needed for these. That solution should be as close to the native browser system as possible, and should leverage its APIs where they are available.

**Comment by kelson42** [comment]
I would recommend to really try to secure we have proper URLs - which is I guess the main problem to get this working with the browser bookmark system. Then re-evaluate how important is that to provide a bookmark system for other use cases. Developing a simple bookmarking system is easy, but having a compeeting feature here will need a lot of work...

**Comment by Aphexus** [comment]
I have installed Kiwi JS for windows and there seems to be no way to bookmark any pages.

it would be nice if there was some way to bookmark hierarchically and even a history. It shouldn't be that hard to add but would be very useful.

The "issue" suggests there are bookmarks for desktop. Since there are no tabs basically one is stuck viewing one page at a time and then no way to really jump around when needed or to resume from previous work.

**Comment by Jaifroid** [comment]
@Aphexus The installed version (Electron or PWA on Chrome/Edge) has access to the File System, so a bookmarking system is feasible, but needs development work. There are challenges: on Firefox and on mobile, there are no persistent permissions, so it would be impossible to open an archive seamlessly (without picking it again) when clicking on a bookmark. If we make use of the Origin Private File System the default, then seamless access is possible on mobile and in Firefox, but unfortunately the quota for Firefox is tiny and not suitable for even medium-size ZIMs (see https://bugzilla.mozilla.org/show_bug.cgi?id=1892994).

Regarding tabs, both the Kiwix Desktop app and Kiwix JS / PWA support multiple tabs or windows for articles in the currently open ZIM (just right-click a link). But are you referring to multiple tabs for different ZIMs being read concurrently? This is possible in Kiwix JS PWA (rather than Electron), which allows opening multiple instances of the app, each with a different archive loaded. Electron currently does not allow multiple instances.

**Comment by Aphexus** [comment]
@Jaifroid Ok. I am relatively new to Kiwi so I'm going to spend some more time playing around with it and trying the different versions. I wasn't aware initially that there was such differences between the desktop version and the js version.

As an aside then, is it possible to restrict the random page to a certain subset of the zim? I have the full wikipedia and when I use random page I get a bunch of things I'm not interested in. E.g., some things I'm not at all interested in like entertainment, sports, etc.

**Comment by Jaifroid** [comment]
Yes, the JS version is a parallel version that has rewritten the C++ backend entirely in JS designed to run in the browser (or as an installed PWA or Electron app, which still uses the browser as the execution platform). It is adapted to the kind of UI that browsers support and JS frameworks like Bootstrap.

To answer your question about the Random button, this is a very simple function which merely takes all the URLs in the ZIM, picks one at random, tests that it is an HTML article, and displays it. If you want a ZIM that is focused on a specific thematic subset of articles, then take a look at the subsets published in https://library.kiwix.org/#lang=eng&category=wikipedia (or select a different language) and browse what's available.

**Comment by Aphexus** [comment]
@Jaifroid  Yes, but couldn't Kiwix simply apply some filter when it chooses a random article? E.g., one, say, creates a regex expression in settings and then as the app chooses a random article it will use the regex to see if it passes or fails and if it fails it then selects another random article until it passes. This would be one way.

The regex could be used on the url + title or even text in the page(in which case it likely would be much faster to do a search and then select randomly out of the search results). Even a simple check on the category would enable limiting a lot of "junk". e.g., a setting "Included/Excluded(+/-) categories for searching: -All; +Math; +Music"

If I do a subset zim it would take a quite a bit of space.

E.g., if I want music + biology + chemistry + mathematics + science + art + history then it is quite a bit of wasted space just to limit the randomization selection.
